src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Android.cs

```diff
@@ -0,0 +1,473 @@
+        {
+            if (!File.Exists(path))
+                return false;
+
+            try
+            {
+                data = File.OpenRead(path);
+            }
+            catch (IOException)
+            {
+                return false;
+            }
+            catch (UnauthorizedAccessException)
+            {
+                return false;
+            }
+
+            try
+            {
+                ReadHeader();
+                return true;
+            }
+            catch
+            {
+                // log something here instead of the console.
+                //Console.Error.WriteLine ("tzdata file \"{0}\" was present but invalid: {1}", path, e);
+            }
+            return false;
+        }
+
+        private unsafe void ReadHeader()
+        {
+            int size   = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+            var buffer = new byte[size];
+            var header = ReadAt<AndroidTzDataHeader>(0, buffer);
+
+            header.indexOffset = NetworkToHostOrder(header.indexOffset);
+            header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+            sbyte* s = (sbyte*)header.signature;
+            string magic = new string(s, 0, 6, Encoding.ASCII);
+
+            if (magic != "tzdata" || header.signature[11] != 0)
+            {
+                var b = new StringBuilder();
+                b.Append("bad tzdata magic:");
+                for (int i = 0; i < 12; ++i) {
+                    b.Append(' ').Append(((byte)s[i]).ToString("x2"));
+                }
+
+                //TODO: Put strings in resource file
+                throw new InvalidOperationException("bad tzdata magic: " + b.ToString());
+            }
+
+            ReadIndex(header.indexOffset, header.dataOffset, buffer);
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2087:UnrecognizedReflectionPattern",
+            Justification = "Implementation detail of Android TimeZone")]
+        private unsafe T ReadAt<T>(long position, byte[] buffer)
+            where T : struct
+        {
+            int size = Marshal.SizeOf(typeof(T));
+            if (buffer.Length < size)
+            {
+                //TODO: Put strings in resource file
+                throw new InvalidOperationException("Internal error: buffer too small");
+            }
+
+            data!.Position = position;
+            int r;
+            if ((r = data!.Read(buffer, 0, size)) < size)
+            {
+                //TODO: Put strings in resource file
+                throw new InvalidOperationException(
+                        string.Format("Error reading '{0}': read {1} bytes, expected {2}", tzdataPath, r, size));
+            }
+
+            fixed(byte* b = buffer)
+            {
+                return (T)Marshal.PtrToStructure((IntPtr)b, typeof(T))!;
+            }
+        }
+
+        private static int NetworkToHostOrder(int value)
```

Comment:
It's strange this logic does not exist already somewhere

Reply:
I'll look for it in the repo.

Reply:
I haven't found it in the pal yet.

Reply:
Should we put it somewhere else?